pkg/operation/botanist/health_check.go

```diff
@@ -16,172 +16,422 @@ package botanist
+	}
+
+	if exitCondition := checkDaemonSets(condition, daemonSetList.Items); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	// TODO: Decide on whether to return own condition for addons / remove addons entirely.
+	return nil, nil
 }
 
-func verifyPodHealthiness(condition *gardenv1beta1.Condition, k8sClient kubernetes.Client, namespace string, listOptions metav1.ListOptions) (*gardenv1beta1.Condition, bool) {
-	podList, err := k8sClient.ListPods(namespace, listOptions)
+func (b *Botanist) checkLoggingControlPlane(condition *gardenv1beta1.Condition) (*gardenv1beta1.Condition, error) {
+	if exitCondition := b.checkShootHibernated(condition); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	deploymentList, err := b.K8sSeedClient.ListDeployments(b.Shoot.SeedNamespace, loggingListOptions)
 	if err != nil {
-		return helper.ModifyCondition(condition, corev1.ConditionUnknown, "FetchPodListFailed", err.Error()), true
-	}
-
-	for _, pod := range podList.Items {
-		for _, containerStatus := range pod.Status.ContainerStatuses {
-			if containerStatus.State.Waiting != nil {
-				return helper.ModifyCondition(condition, corev1.ConditionFalse, "ContainerWaiting", fmt.Sprintf("Container %s of pod %s is waiting to start.", containerStatus.Name, pod.Name)), true
-			}
-			if containerStatus.State.Running != nil && !containerStatus.Ready {
-				return helper.ModifyCondition(condition, corev1.ConditionFalse, "ContainerNotReady", fmt.Sprintf("Container %s of pod %s is not ready yet.", containerStatus.Name, pod.Name)), true
-			}
-			if containerStatus.State.Terminated != nil && containerStatus.State.Terminated.Reason != "Completed" {
-				return helper.ModifyCondition(condition, corev1.ConditionFalse, "ContainerNotRunning", fmt.Sprintf("Container %s of pod %s is not in running state.", containerStatus.Name, pod.Name)), true
-			}
-		}
-		if pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodSucceeded {
-			return helper.ModifyCondition(condition, corev1.ConditionFalse, "PodNotRunning", fmt.Sprintf("Pod %s is in phase %s", pod.Name, pod.Status.Phase)), true
-		}
+		return nil, err
+	}
+
+	if exitCondition := checkRequiredDeployments(condition, common.RequiredLoggingDeployments, deploymentList.Items); exitCondition != nil {
+		return exitCondition, nil
+	}
+	if exitCondition := checkDeployments(condition, deploymentList.Items); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	statefulSetList, err := b.K8sSeedClient.ListStatefulSets(b.Shoot.SeedNamespace, loggingListOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if exitCondition := checkRequiredStatefulSets(condition, common.RequiredLoggingStatefulSets, statefulSetList.Items); exitCondition != nil {
+		return exitCondition, nil
+	}
+	if exitCondition := checkStatefulSets(condition, statefulSetList.Items); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	daemonSetList, err := b.K8sSeedClient.ListDaemonSets(b.Shoot.SeedNamespace, loggingListOptions)
```

Comment:
I see a lot of `List{Deployments,StatefulSets,DaemonSets}` calls in this file (though, with different list options/label selectors, but still). Can we please reduce this to exactly ONE `ListDeployments`, `ListStatefulSets`, `ListDaemonSets` call and filter ourselves on the wanted object in the individual functions?